src/SpirvIR.cpp

```diff
@@ -596,120 +703,343 @@ void SpvModule::encode(SpvBinary &binary) const {
+    }
+    };
+    return "InvalidItem";
+}
+
+SpvKind SpvBuilder::kind_of(SpvId item_id) const {
     KindMap::const_iterator it = kind_map.find(item_id);
     if (it != kind_map.end()) {
-        return SpvInvalidItem;
+        return it->second;
+    }
+    return SpvInvalidItem;
+}
+
+SpvId SpvBuilder::type_of(SpvId variable_id) const {
+    VariableTypeMap::const_iterator it = variable_type_map.find(variable_id);
+    if (it != variable_type_map.end()) {
+        return it->second;
+    }
+    return SpvInvalidId;
+}
+
+void SpvBuilder::finalize() {
+    SpvId last_id = (SpvId)(kind_map.size() + 1);
+    module.set_binding_count(last_id);
+
+    if (module.is_capability_required(SpvCapabilityInt8)) {
+        module.require_extension("SPV_KHR_8bit_storage");
+    }
+
+    if (module.is_capability_required(SpvCapabilityInt16)) {
+        module.require_extension("SPV_KHR_16bit_storage");
     }
-    return it->second;
 }
 
 void SpvBuilder::encode(SpvBinary &binary) const {
     // Encode the module
     module.encode(binary);
 }
 
-SpvId SpvBuilder::map_type(const Type &type, uint32_t array_size) {
+SpvId SpvBuilder::declare_type(const Type &type, uint32_t array_size) {
     SpvId type_id = lookup_type(type, array_size);
     if (type_id == SpvInvalidId) {
-        type_id = declare_type(type, array_size);
+        type_id = add_type(type, array_size);
     }
     return type_id;
 }
 
-SpvId SpvBuilder::map_pointer_type(const Type &type, SpvStorageClass storage_class) {
+SpvId SpvBuilder::declare_pointer_type(const Type &type, SpvStorageClass storage_class) {
     SpvId ptr_type_id = lookup_pointer_type(type, storage_class);
     if (ptr_type_id == SpvInvalidId) {
-        ptr_type_id = declare_pointer_type(ptr_type_id, storage_class);
+        ptr_type_id = add_pointer_type(ptr_type_id, storage_class);
     }
     return ptr_type_id;
 }
 
-SpvId SpvBuilder::map_pointer_type(SpvId type_id, SpvStorageClass storage_class) {
+SpvId SpvBuilder::declare_pointer_type(SpvId type_id, SpvStorageClass storage_class) {
     SpvId ptr_type_id = lookup_pointer_type(type_id, storage_class);
     if (ptr_type_id == SpvInvalidId) {
-        ptr_type_id = declare_pointer_type(type_id, storage_class);
+        ptr_type_id = add_pointer_type(type_id, storage_class);
     }
     return ptr_type_id;
 }
 
-SpvId SpvBuilder::map_function_type(SpvId return_type, const ParamTypes &param_types) {
+SpvId SpvBuilder::declare_function_type(SpvId return_type, const ParamTypes &param_types) {
     SpvId type_id = lookup_function_type(return_type, param_types);
     if (type_id == SpvInvalidId) {
-        type_id = declare_function_type(return_type, param_types);
+        type_id = add_function_type(return_type, param_types);
     }
     return type_id;
 }
 
-SpvId SpvBuilder::map_constant(const Type &type, const void *data) {
+SpvId SpvBuilder::declare_function(const std::string &name, SpvId function_type) {
+    SpvId existing_id = lookup_id(name);
```

Comment:
Can you add a comment as to why re-declaring a function isn't an error?  I'm a bit confused why we need this logic here.

Reply:
Sure thing ... I'll add doc-strings for the builder class.  Yeah, these declaration methods are really convenience methods for "if this thing isn't defined already, then define it, otherwise give me its existing id", which avoids making the caller do all these checks, and avoids inadvertently creating duplicates.

Reply:
Added!